### PR TITLE
feat: add option `allowEscape` to `no-misleading-character-class` rule

### LINKED ISSUE #15080: Bug: no-misleading-character-class triggers on seemingly-OK code
### Environment

Node version: 16.9.1
npm version: 7.21.1
Local ESLint version: 7.32.0
Global ESLint version: (none)
Operating System: Windows 10

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

Just using eslint:recommended. This can be [demoed on eslint.org](https://eslint.org/demo#eyJ0ZXh0IjoiL1tcXHUwMEI3XFx1MDMwMC1cXHUwMzZGXS91LnRlc3QoXCJcIik7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)
</details>

```js
/[\u00B7\u0300-\u036F]/u.test("");
```

### What did you expect to happen?

No error, if I am understanding https://eslint.org/docs/rules/no-misleading-character-class correctly. Those docs imply the error will trigger when I include in my source text a multi-code-point character, which I have not done; I've only used Unicode escapes.

### What actually happened?

```
4:3   error  Unexpected combined character in character class  no-misleading-character-class
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

For context, I am trying to reproduce https://www.w3.org/TR/REC-xml/#NT-NameStartChar as a RegExp. The `#xB7 | [#x0300-#x036F]` part of the grammar is what maps to my `[\u00B7\u0300-\u036F]` character class.

It's possible that this is a doc issue, as I'm actually doing something dangerous or unusual, and so the error should trigger? But if I am, it's not clear at all from the docs what I'm doing that's bad.

### PR DESCRIPTION
**Pull request by fasttime**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Fixes #15080 

#### What changes did you make? (Give an overview)

This PR adds an option to allow escape sequences in regexp character classes to the `no-misleading-character-class` rule. The new option only has effect on nodes whose source location can be statically determined with our tools. These are the same nodes that report granular errors:
* regular expression literals
* string literals as first arguments of the `RegExp` constructor
* template literals without expressions as first arguments of the `RegExp` constructor

#### Is there anything you'd like reviewers to focus on?

I would like to fine tune this rule to allow only escape sequence that are not confusing, and I'm not sure I've got the right balance. For example `/[\👍]/` contains an escape sequence where the backslash just escapes the next character without changing its value. I think that those escape sequences should be still forbidden even when the new option is set. Other cases are less trivial.

In the current implementation, the new option will allow any escape sequences except those where the character being escaped is prepended by one or more backslashes. If one or more of the characters in a group are escaped, the group will be allowed.

| Node | Status |
|------|--------|
| `/[A\u0301]/` | allowed, second character in class is escaped |
| `/[👨\u200d👩\u200d👦]/u` | allowed, ZWNJ character is escaped |
| `/[\n̅]/` | allowed, `\n` is a non-trivial escape sequence |
| `/[\e̅]/` | not allowed, `\e` is `e` |

### CODE REVIEW

**Comment by mdjermanovic** [comment]
> `/[\n̅]/`	allowed, `\n` is a non-trivial escape sequence
> `/[\e̅]/`	not allowed, `\e` is `e`

These two look like the same kind of possible error this rule targets - the sequence appears as a single character - so perhaps both should not be allowed?

**Comment by fasttime** [comment]
> > `/[\n̅]/`	allowed, `\n` is a non-trivial escape sequence
> > `/[\e̅]/`	not allowed, `\e` is `e`
> 
> These two look like the same kind of possible error this rule targets - the sequence appears as a single character - so perhaps both should not be allowed?

Maybe we could require that both characters in `n̅` are escaped? I'm not sure, because the problem here seems to lie in the unescaped combining character itself rather than the grouping of characters, and the rule does not flag those characters everywhere. This means that character classes like `/[̳]/` or `/[\n\u0305̳abc]/` will remain allowed despite looking confusing.

**Comment by fasttime** [comment]
Perhaps we could allow only sequences that consist exclusively of ASCII characters. So `/[\n\u0305]/` would be allowed, but `/[\n̅]/` or `/[\n\̅]/` would be not. What do you think @mdjermanovic?

**Comment by fasttime** [comment]
I think we can require that combining mark characters in a character class are only allowed when written as escape sequences; which is to say regardless of the previous character. This would be the most immediate solution to always disallow `/[\n̅]/` and `/[\u006e̅]/` along with `/[\e̅]/`. I'll work on an update.

**Comment by nzakas** [comment]
@mdjermanovic looking for your feedback here.

**Review by mdjermanovic — APPROVED** [review]
LGTM. Great work, thanks!

Since this is a new feature, it needs another approval from a team member before merging.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
